File: 2001-3000/2330. Valid Palindrome IV.java
// Medium
// Two Pointers
// Time:O(n), Space:O(1)
// https://leetcode.cn/problems/valid-palindrome-iv/

class Solution {
    public boolean makePalindrome(String s) { 
        int left = 0, right = s.length() - 1, cnt = 0;
        while (left < right) {
            if (s.charAt(left) != s.charAt(right)) {
                cnt++;
            }
            left++;
            right--;
        }
        return cnt <= 2;
    }
}

File: 2001-3000/2330. Valid Palindrome IV.py
# Time:O(n), Space:O(1)

class Solution:
    def makePalindrome(self, s: str) -> bool:
        left, right = 0, len(s) - 1
        mismatch = 0

        while left < right:
            if s[left] != s[right]:
                mismatch += 1
                if mismatch > 2:
                    return False
            left += 1
            right -= 1

        return True
